pkg/util/crypto.go

```diff
@@ -31,34 +32,103 @@ import (
+		return err
+	}
+
+	template := x509.Certificate{
+		SerialNumber: big.NewInt(1),
+		Subject: pkix.Name{
+			CommonName: "minikubeCA",
+		},
+		NotBefore: time.Now(),
+		NotAfter:  time.Now().Add(time.Hour * 24 * 365 * 10),
+
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+		IsCA: true,
+	}
+
+	return writeCertsAndKeys(&template, certPath, priv, keyPath, &template, priv)
+}
+
 // You may also specify additional subject alt names (either ip or dns names) for the certificate
 // The certificate will be created with file mode 0644. The key will be created with file mode 0600.
 // If the certificate or key files already exist, they will be overwritten.
 // Any parent directories of the certPath or keyPath will be created as needed with file mode 0755.
-func GenerateSelfSignedCert(certPath, keyPath string, ips []net.IP, alternateDNS []string) error {
-	priv, err := rsa.GenerateKey(rand.Reader, 2048)
+func GenerateSignedCert(certPath, keyPath string, ips []net.IP, alternateDNS []string, signerCertPath, signerKeyPath string) error {
+	signerCertBytes, err := ioutil.ReadFile(signerCertPath)
+	if err != nil {
+		return err
+	}
+	decodedSignerCert, _ := pem.Decode(signerCertBytes)
+	if decodedSignerCert == nil {
+		return fmt.Errorf("Unable to decode certificate.")
+	}
+	signerCert, err := x509.ParseCertificate(decodedSignerCert.Bytes)
+	if err != nil {
+		return err
+	}
+	signerKeyBytes, err := ioutil.ReadFile(signerKeyPath)
+	if err != nil {
+		return err
+	}
+	decodedSignerKey, _ := pem.Decode(signerKeyBytes)
+	if decodedSignerKey == nil {
+		return fmt.Errorf("Unable to decode key.")
+	}
+	signerKey, err := x509.ParsePKCS1PrivateKey(decodedSignerKey.Bytes)
 	if err != nil {
 		return err
 	}
 
 	template := x509.Certificate{
-		SerialNumber: big.NewInt(1),
+		SerialNumber: big.NewInt(2),
 		Subject: pkix.Name{
 			CommonName: "minikube",
 		},
 		NotBefore: time.Now(),
 		NotAfter:  time.Now().Add(time.Hour * 24 * 365),
 
-		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
 		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth},
 		BasicConstraintsValid: true,
-		IsCA: true,
 	}
 
 	template.IPAddresses = append(template.IPAddresses, ips...)
 	template.DNSNames = append(template.DNSNames, alternateDNS...)
 
-	derBytes, err := x509.CreateCertificate(rand.Reader, &template, &template, &priv.PublicKey, priv)
+	priv, err := loadOrGeneratePrivateKey(keyPath)
+	if err != nil {
+		return err
+	}
+
+	return writeCertsAndKeys(&template, certPath, priv, keyPath, signerCert, signerKey)
+}
+
+func loadOrGeneratePrivateKey(keyPath string) (*rsa.PrivateKey, error) {
```

Comment:
With a long-lived CA cert, do we need to attempt to re-use the private key still?